File: www/cgi/python/calc.py
#!/usr/bin/env python3
import os
import sys
import urllib.parse

def parse_post_data():
    """Parse POST data from stdin"""
    content_length = int(os.environ.get('CONTENT_LENGTH', 0))
    if content_length > 0:
        post_data = sys.stdin.read(content_length)
        return urllib.parse.parse_qs(post_data)
    return {}

def calculate(num1, num2, operation):
    """Perform the calculation based on operation"""
    try:
        n1 = float(num1)
        n2 = float(num2)
        
        operations = {
            'add': (n1 + n2, '+'),
            'subtract': (n1 - n2, '-'),
            'multiply': (n1 * n2, '×'),
            'divide': (n1 / n2 if n2 != 0 else None, '÷'),
            'power': (n1 ** n2, '^'),
            'modulo': (n1 % n2 if n2 != 0 else None, '%')
        }
        
        if operation not in operations:
            return None, None, "Invalid operation"
        
        result, symbol = operations[operation]
        
        if result is None:
            return None, symbol, "Division by zero"
        
        return result, symbol, None
        
    except ValueError:
        return None, None, "Invalid number format"
    except Exception as e:
        return None, None, str(e)

def main():
    # Parse POST data
    data = parse_post_data()
    
    # Extract parameters
    num1 = data.get('num1', [''])[0]
    num2 = data.get('num2', [''])[0]
    operation = data.get('operation', [''])[0]
    
    # Send HTTP headers
    print("Content-Type: text/html\r")
    print("\r")
    
    # Validate input
    if not num1 or not num2 or not operation:
        print("<h3>❌ Error</h3>")
        print("<p>Missing parameters. Please provide num1, num2, and operation.</p>")
        return
    
    # Calculate
    result, symbol, error = calculate(num1, num2, operation)
    
    if error:
        print("<h3>❌ Error</h3>")
        print(f"<p>{error}</p>")
    else:
        print("<h3>✅ Result</h3>")
        print(f"<div class='result-value'>{num1} {symbol} {num2} = {result}</div>")
        print(f"<p><strong>Operation:</strong> {operation.capitalize()}</p>")
        print(f"<p><strong>Language:</strong> Python 3 🐍</p>")

if __name__ == "__main__":
    main()

File: www/cgi/php/calc.php
#!/usr/bin/php-cgi
<?php

function parsePostData() {
    // Try to get CONTENT_LENGTH from environment
    $contentLength = getenv('CONTENT_LENGTH');
    if ($contentLength === false || $contentLength == '') {
        $contentLength = 0;
    } else {
        $contentLength = intval($contentLength);
    }
    
    if ($contentLength > 0) {
        $stdin = fopen('php://stdin', 'r');
        $postData = fread($stdin, $contentLength);
        fclose($stdin);
        parse_str($postData, $data);
        return $data;
    }
    return array();
}

function calculate($num1, $num2, $operation) {
    $n1 = floatval($num1);
    $n2 = floatval($num2);
    
    $operations = array(
        'add' => array('result' => $n1 + $n2, 'symbol' => '+'),
        'subtract' => array('result' => $n1 - $n2, 'symbol' => '-'),
        'multiply' => array('result' => $n1 * $n2, 'symbol' => '×'),
        'divide' => array('result' => $n2 != 0 ? $n1 / $n2 : null, 'symbol' => '÷'),
        'power' => array('result' => pow($n1, $n2), 'symbol' => '^'),
        'modulo' => array('result' => $n2 != 0 ? fmod($n1, $n2) : null, 'symbol' => '%')
    );
    
    if (!array_key_exists($operation, $operations)) {
        return array('error' => 'Invalid operation', 'symbol' => null, 'result' => null);
    }
    
    $op = $operations[$operation];
    
    if ($op['result'] === null) {
        return array('error' => 'Division by zero', 'symbol' => $op['symbol'], 'result' => null);
    }
    
    return array('result' => $op['result'], 'symbol' => $op['symbol'], 'error' => null);
}

// Send HTTP headers first
echo "Content-Type: text/html\r\n";
echo "\r\n";

// Parse POST data
$data = parsePostData();

// Extract parameters
$num1 = isset($data['num1']) ? $data['num1'] : '';
$num2 = isset($data['num2']) ? $data['num2'] : '';
$operation = isset($data['operation']) ? $data['operation'] : '';

// Validate input
if (empty($num1) || empty($num2) || empty($operation)) {
    echo "<h3>❌ Error</h3>";
    echo "<p>Missing parameters. Please provide num1, num2, and operation.</p>";
    exit;
}

// Calculate
$calc = calculate($num1, $num2, $operation);

if ($calc['error']) {
    echo "<h3>❌ Error</h3>";
    echo "<p>" . htmlspecialchars($calc['error']) . "</p>";
} else {
    echo "<h3>✅ Result</h3>";
    echo "<div class='result-value'>" . htmlspecialchars($num1) . " " . htmlspecialchars($calc['symbol']) . " " . htmlspecialchars($num2) . " = " . htmlspecialchars($calc['result']) . "</div>";
    echo "<p><strong>Operation:</strong> " . htmlspecialchars(ucfirst($operation)) . "</p>";
    echo "<p><strong>Language:</strong> PHP 🐘</p>";
}
?>
